<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDA0321ENSkillsNetwork928-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="400" alt="Skills Network Logo"  />
    </a>
</p>


# **Data Wrangling Lab**


Estimated time needed: **45 to 60** minutes


In this assignment you will be performing data wrangling.


## Objectives


In this lab you will perform the following:


-   Identify duplicate values in the dataset.

-   Remove duplicate values from the dataset.

-   Identify missing values in the dataset.

-   Impute the missing values in the dataset.

-   Normalize data in the dataset.


<hr>


## Hands on Lab


Import pandas module.


In [3]:
%pip install pandas

In [4]:
import pandas as pd

Load the dataset into a dataframe.


<h2>Read Data</h2>
<p>
We utilize the <code>pandas.read_csv()</code> function for reading CSV files. However, in this version of the lab, which operates on JupyterLite, the dataset needs to be downloaded to the interface using the provided code below.
</p>


The functions below will download the dataset into your browser:


In [5]:
from pyodide.http import pyfetch

async def download(url, filename):
    response = await pyfetch(url)
    if response.status == 200:
        with open(filename, "wb") as f:
            f.write(await response.bytes())

In [6]:
file_path = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DA0321EN-SkillsNetwork/LargeData/m1_survey_data.csv"

To obtain the dataset, utilize the download() function as defined above:  


In [7]:
await download(file_path, "m1_survey_data.csv")
file_name="m1_survey_data.csv"

Utilize the Pandas method read_csv() to load the data into a dataframe.


In [8]:
df = pd.read_csv(file_name)

> Note: This version of the lab is working on JupyterLite, which requires the dataset to be downloaded to the interface.While working on the downloaded version of this notebook on their local machines(Jupyter Anaconda), the learners can simply **skip the steps above,** and simply use the URL directly in the `pandas.read_csv()` function. You can uncomment and run the statements in the cell below.


In [ ]:
#df = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DA0321EN-SkillsNetwork/LargeData/m1_survey_data.csv")

## Finding duplicates


In this section you will identify duplicate values in the dataset.


 Find how many duplicate rows exist in the dataframe.


In [45]:
# your code goes here
df.duplicated().sum()


0

## Removing duplicates


Remove the duplicate rows from the dataframe.


In [18]:
# your code goes here
df.drop_duplicates(inplace = True)

Verify if duplicates were actually dropped.


In [20]:
# your code goes here
dups = df.duplicated(subset = ['Respondent'])
print(df[dups])
df.describe()

Empty DataFrame
Columns: [Respondent, MainBranch, Hobbyist, OpenSourcer, OpenSource, Employment, Country, Student, EdLevel, UndergradMajor, EduOther, OrgSize, DevType, YearsCode, Age1stCode, YearsCodePro, CareerSat, JobSat, MgrIdiot, MgrMoney, MgrWant, JobSeek, LastHireDate, LastInt, FizzBuzz, JobFactors, ResumeUpdate, CurrencySymbol, CurrencyDesc, CompTotal, CompFreq, ConvertedComp, WorkWeekHrs, WorkPlan, WorkChallenge, WorkRemote, WorkLoc, ImpSyn, CodeRev, CodeRevHrs, UnitTests, PurchaseHow, PurchaseWhat, LanguageWorkedWith, LanguageDesireNextYear, DatabaseWorkedWith, DatabaseDesireNextYear, PlatformWorkedWith, PlatformDesireNextYear, WebFrameWorkedWith, WebFrameDesireNextYear, MiscTechWorkedWith, MiscTechDesireNextYear, DevEnviron, OpSys, Containers, BlockchainOrg, BlockchainIs, BetterLife, ITperson, OffOn, SocialMedia, Extraversion, ScreenName, SOVisit1st, SOVisitFreq, SOVisitTo, SOFindAnswer, SOTimeSaved, SOHowMuchTime, SOAccount, SOPartFreq, SOJobs, EntTeams, SOComm, WelcomeChang

Respondent     CompTotal  ConvertedComp   WorkWeekHrs   CodeRevHrs  \
count  11398.000000  1.058900e+04   1.058200e+04  11276.000000  8972.000000   
mean   12490.392437  7.570477e+05   1.315967e+05     42.064606     4.781071   
std     7235.461999  9.705598e+06   2.947865e+05     24.672741     4.567060   
min        4.000000  0.000000e+00   0.000000e+00      3.000000     0.000000   
25%     6264.250000  2.500000e+04   2.686800e+04     40.000000     2.000000   
50%    12484.000000  6.500000e+04   5.774500e+04     40.000000     4.000000   
75%    18784.750000  1.200000e+05   1.000000e+05     43.000000     5.000000   
max    25142.000000  7.000000e+08   2.000000e+06   1012.000000    99.000000   

                Age  
count  11111.000000  
mean      30.778895  
std        7.393686  
min       16.000000  
25%       25.000000  
50%       29.000000  
75%       35.000000  
max       99.000000

## Finding Missing values


Find the missing values for all columns.


In [22]:
# your code goes here
missing_data = df.isnull()
print(missing_data.head())

   Respondent  MainBranch  Hobbyist  OpenSourcer  OpenSource  Employment  \
0       False       False     False        False       False       False   
1       False       False     False        False       False       False   
2       False       False     False        False       False       False   
3       False       False     False        False       False       False   
4       False       False     False        False       False       False   

   Country  Student  EdLevel  UndergradMajor  ...  WelcomeChange  \
0    False    False    False           False  ...          False   
1    False    False    False           False  ...          False   
2    False    False    False           False  ...          False   
3    False    False    False            True  ...          False   
4    False    False    False           False  ...          False   

   SONewContent    Age  Gender  Trans  Sexuality  Ethnicity  Dependents  \
0         False  False   False  False      False      False

In [23]:
for column in missing_data.columns.values.tolist():
    print(column)
    print(missing_data[column].value_counts())
    print("")



Respondent
Respondent
False    11398
Name: count, dtype: int64

MainBranch
MainBranch
False    11398
Name: count, dtype: int64

Hobbyist
Hobbyist
False    11398
Name: count, dtype: int64

OpenSourcer
OpenSourcer
False    11398
Name: count, dtype: int64

OpenSource
OpenSource
False    11317
True        81
Name: count, dtype: int64

Employment
Employment
False    11398
Name: count, dtype: int64

Country
Country
False    11398
Name: count, dtype: int64

Student
Student
False    11347
True        51
Name: count, dtype: int64

EdLevel
EdLevel
False    11286
True       112
Name: count, dtype: int64

UndergradMajor
UndergradMajor
False    10661
True       737
Name: count, dtype: int64

EduOther
EduOther
False    11234
True       164
Name: count, dtype: int64

OrgSize
OrgSize
False    11302
True        96
Name: count, dtype: int64

DevType
DevType
False    11333
True        65
Name: count, dtype: int64

YearsCode
YearsCode
False    11389
True         9
Name: count, dtype: int64

Age1stCode
Age

Find out how many rows are missing in the column 'WorkLoc'


In [27]:
# your code goes here
missing_rows = df['Respondent'].count() - df['WorkLoc'].count()
print('There are {} missing rows in the WorkLoc column'.format(missing_rows))

There are 32 missing rows in the WorkLoc column


## Imputing missing values


Find the  value counts for the column WorkLoc.


In [25]:
# your code goes here
print(missing_data['WorkLoc'].value_counts())

WorkLoc
False    11366
True        32
Name: count, dtype: int64


Identify the value that is most frequent (majority) in the WorkLoc column.


In [29]:
#make a note of the majority value here, for future reference
most_freq = df['WorkLoc'].value_counts().idxmax()
print('The most frequent value in the WorkLoc column is: {}'.format(most_freq))

The most frequent value in the WorkLoc column is: Office


Impute (replace) all the empty rows in the column WorkLoc with the value that you have identified as majority.


In [35]:
# your code goes here
df['WorkLoc'] = df['WorkLoc'].fillna('Office')

After imputation there should ideally not be any empty rows in the WorkLoc column.


Verify if imputing was successful.


In [36]:
# your code goes here
print(df['WorkLoc'].value_counts())

WorkLoc
Office                                            6838
Home                                              3589
Other place, such as a coworking space or cafe     971
Name: count, dtype: int64


## Normalizing data


There are two columns in the dataset that talk about compensation.

One is "CompFreq". This column shows how often a developer is paid (Yearly, Monthly, Weekly).

The other is "CompTotal". This column talks about how much the developer is paid per Year, Month, or Week depending upon his/her "CompFreq". 

This makes it difficult to compare the total compensation of the developers.

In this section you will create a new column called 'NormalizedAnnualCompensation' which contains the 'Annual Compensation' irrespective of the 'CompFreq'.

Once this column is ready, it makes comparison of salaries easy.


<hr>


List out the various categories in the column 'CompFreq'


In [37]:
# your code goes here
df['CompFreq'].value_counts()

CompFreq
Yearly     6073
Monthly    4788
Weekly      331
Name: count, dtype: int64

Create a new column named 'NormalizedAnnualCompensation'. Use the hint given below if needed.


In [39]:
df['NormalizedAnnualCompensation'] = ''
df.columns

Index(['Respondent', 'MainBranch', 'Hobbyist', 'OpenSourcer', 'OpenSource',
       'Employment', 'Country', 'Student', 'EdLevel', 'UndergradMajor',
       'EduOther', 'OrgSize', 'DevType', 'YearsCode', 'Age1stCode',
       'YearsCodePro', 'CareerSat', 'JobSat', 'MgrIdiot', 'MgrMoney',
       'MgrWant', 'JobSeek', 'LastHireDate', 'LastInt', 'FizzBuzz',
       'JobFactors', 'ResumeUpdate', 'CurrencySymbol', 'CurrencyDesc',
       'CompTotal', 'CompFreq', 'ConvertedComp', 'WorkWeekHrs', 'WorkPlan',
       'WorkChallenge', 'WorkRemote', 'WorkLoc', 'ImpSyn', 'CodeRev',
       'CodeRevHrs', 'UnitTests', 'PurchaseHow', 'PurchaseWhat',
       'LanguageWorkedWith', 'LanguageDesireNextYear', 'DatabaseWorkedWith',
       'DatabaseDesireNextYear', 'PlatformWorkedWith',
       'PlatformDesireNextYear', 'WebFrameWorkedWith',
       'WebFrameDesireNextYear', 'MiscTechWorkedWith',
       'MiscTechDesireNextYear', 'DevEnviron', 'OpSys', 'Containers',
       'BlockchainOrg', 'BlockchainIs', 'BetterLife'

Double click to see the **Hint**.

<!--

Use the below logic to arrive at the values for the column NormalizedAnnualCompensation.

If the CompFreq is Yearly then use the exising value in CompTotal
If the CompFreq is Monthly then multiply the value in CompTotal with 12 (months in an year)
If the CompFreq is Weekly then multiply the value in CompTotal with 52 (weeks in an year)

-->


In [48]:
# your code goes here
def convert(n):
    if (n['CompFreq'] == 'Weekly'):
        return n['CompTotal'] * 52
    elif (n['CompFreq'] == 'Monthly'):
       return n['CompTotal'] * 12
    else:
      return n['CompTotal']

df['NormalizedAnnualCompensation'] = df.apply(convert, axis = 1)        

In [51]:
df[['CompFreq', 'CompTotal', 'NormalizedAnnualCompensation']].head(21)

CompFreq  CompTotal  NormalizedAnnualCompensation
0    Yearly    61000.0                       61000.0
1    Yearly   138000.0                      138000.0
2    Yearly    90000.0                       90000.0
3   Monthly    29000.0                      348000.0
4    Yearly    90000.0                       90000.0
5   Monthly     9500.0                      114000.0
6   Monthly     3000.0                       36000.0
7    Yearly   103000.0                      103000.0
8    Yearly    69000.0                       69000.0
9   Monthly     8000.0                       96000.0
10  Monthly     7000.0                       84000.0
11   Yearly   114000.0                      114000.0
12   Weekly     2000.0                      104000.0
13   Weekly    22000.0                     1144000.0
14  Monthly    96000.0                     1152000.0
15   Yearly   156000.0                      156000.0
16   Yearly    18000.0                       18000.0
17  Monthly     6400.0                       76800.0
18  Monthly     5000.0                       60000.0
19   Yearly   400000.0                      400000.0
20   Yearly    47300.0                       47300.0

## Authors


Ramesh Sannareddy


### Other Contributors


Rav Ahuja


 ## Change Log


|Date (YYYY-MM-DD)|Version|Changed By|Change Description|
|-|-|-|-|
|2024-09-24|1.1|Madhusudhan Moole|Updated lab|
|2024-09-23|1.0|Raghul Ramesh|Created lab|


<!--| Date (YYYY-MM-DD) | Version | Changed By        | Change Description                 |
| ----------------- | ------- | ----------------- | ---------------------------------- |
| 2020-10-17        | 0.1     | Ramesh Sannareddy | Created initial version of the lab |--!>


## <h3 align="center"> © IBM Corporation. All rights reserved. <h3/>
